src/txHashBuilder.c

```diff
@@ -288,20 +292,358 @@ void txHashBuilder_addCertificate_delegation(
+		{
+			BUILDER_APPEND_CBOR(CBOR_TYPE_UNSIGNED, 0);
+		}
+		{
+			if (port != NULL) {
+				BUILDER_APPEND_CBOR(CBOR_TYPE_UNSIGNED, *port);
+			} else {
+				BUILDER_APPEND_CBOR(CBOR_TYPE_NULL, 0);
+			}
+		}
+		{
+			if (ipv4 != NULL) {
+				ASSERT(ipv4Size == 4);
+				BUILDER_APPEND_CBOR(CBOR_TYPE_BYTES, ipv4Size);
+				BUILDER_APPEND_DATA(ipv4, ipv4Size);
+			} else {
+				BUILDER_APPEND_CBOR(CBOR_TYPE_NULL, 0);
+			}
+		}
+		{
+			if (ipv6 != NULL) {
+				ASSERT(ipv6Size == 16);
+				BUILDER_APPEND_CBOR(CBOR_TYPE_BYTES, ipv6Size);
+
+				// serialized as 4 big-endian uint32
+				for (size_t i = 0; i < 4; i++) {
+					for (size_t j = 0; j < 4; j++) {
+						size_t index = 4 * i + (3 - j);
+						ASSERT(index < ipv6Size);
+
+						BUILDER_APPEND_DATA(ipv6 + index, 1);
+					}
+				}
+			} else {
+				BUILDER_APPEND_CBOR(CBOR_TYPE_NULL, 0);
+			}
+		}
+	}
+}
+
+void txHashBuilder_addPoolRegistrationCertificate_addRelay1(
+        tx_hash_builder_t* builder,
+        const uint16_t* port,
+        const uint8_t* dnsName, size_t dnsNameSize
+)
+{
+	TRACE("txHashBuilder_addPoolRegistrationCertificate_addRelay1: %d", builder->state);
+
+	addRelay_updateState(builder);
+	ASSERT(builder->state == TX_HASH_BUILDER_IN_CERTIFICATES_POOL_RELAYS);
+	ASSERT(builder->poolCertificateData.remainingRelays > 0);
+	builder->poolCertificateData.remainingRelays--;
+
+	ASSERT(dnsName != NULL);
+	ASSERT(dnsNameSize > 0);
+	ASSERT(dnsNameSize < BUFFER_SIZE_PARANOIA);
+
+	// Array(3)[
+	//   Unsigned[1]
+	//   Unsigned[port] / Null
+	//   Text[dnsName]
+	// ]
+	{
+		BUILDER_APPEND_CBOR(CBOR_TYPE_ARRAY, 3);
+		{
+			BUILDER_APPEND_CBOR(CBOR_TYPE_UNSIGNED, 1);
+		}
+		{
+			if (port != NULL) {
+				BUILDER_APPEND_CBOR(CBOR_TYPE_UNSIGNED, *port);
+			} else {
+				BUILDER_APPEND_CBOR(CBOR_TYPE_NULL, 0);
+			}
+		}
+		{
+			BUILDER_APPEND_CBOR(CBOR_TYPE_TEXT, dnsNameSize);
+			BUILDER_APPEND_DATA(dnsName, dnsNameSize);
+		}
+	}
+}
+
+void txHashBuilder_addPoolRegistrationCertificate_addRelay2(
+        tx_hash_builder_t* builder,
+        const uint8_t* dnsName, size_t dnsNameSize
+)
+{
+	TRACE("txHashBuilder_addPoolRegistrationCertificate_addRelay2: %d", builder->state);
+
+	addRelay_updateState(builder);
+	ASSERT(builder->state == TX_HASH_BUILDER_IN_CERTIFICATES_POOL_RELAYS);
+	ASSERT(builder->poolCertificateData.remainingRelays > 0);
+	builder->poolCertificateData.remainingRelays--;
+
+	ASSERT(dnsName != NULL);
+	ASSERT(dnsNameSize > 0);
+	ASSERT(dnsNameSize <= 64);
+
+	// Array(2)[
+	//   Unsigned[2]
+	//   Text[dnsName]
+	// ]
+	{
+		BUILDER_APPEND_CBOR(CBOR_TYPE_ARRAY, 2);
+		{
+			BUILDER_APPEND_CBOR(CBOR_TYPE_UNSIGNED, 2);
+		}
+		{
+			BUILDER_APPEND_CBOR(CBOR_TYPE_TEXT, dnsNameSize);
+			BUILDER_APPEND_DATA(dnsName, dnsNameSize);
+		}
+	}
+}
+
+// enter empty owners or relays if none were received
+static void addPoolMetadata_updateState(tx_hash_builder_t* builder)
+{
+	switch (builder->state) {
+	case TX_HASH_BUILDER_IN_CERTIFICATES_POOL:
```

Comment:
thouldn't there be enter pool owners?

Reply:
Fixed in f8a3823e9908541efd76425fa691875a4867ed1a.